README.md
# Netlify Deploy 🚗💨

<p>
  <img src="https://img.shields.io/badge/version-v0.0.6-blue.svg?cacheSeconds=2592000" />
  <a href="https://github.com/pixelcollective/netlify-deploy/blob/master/LICENSE">
    <img alt="License: MIT" src="https://img.shields.io/badge/License-MIT-yellow.svg" target="_blank" />
  </a>
  <a href="https://packagist.org/pixelcollective/netlify-deploy">
    <img src="https://img.shields.io/packagist/dt/pixelcollective/netlify-deploy.svg?color=purple-blue&style=flat-square" />
  </a>
  <a href="https://github.com/pixelcollective/netlify-deploy/tree/master/CHANGELOG.md">
    <img src="https://img.shields.io/badge/Changelog-Trying-lightgrey.svg" />
  </a>
  <a href="https://twitter.com/tinydevteam">
    <img alt="Twitter: @tinydevteam" src="https://img.shields.io/twitter/follow/tinydevteam.svg?style=social" target="_blank" />
  </a>
</p>

> Automatic Netlify builds on WordPress publish and update events.

## Features

- Makes it super easy to keep a static frontend in sync with your post database
- Out of the box support for configuration with .env
- Full composer support
- Supports custom posttypes
- Supports custom publish hooks
- Free &amp; open source

## Requirements

- WordPress 5.2+
- PHP 7.2+

## Install

```sh
composer require pixelcollective\netlify-deploy
```

## Usage

Request a webhook URL from Netlify to use to trigger builds (you can find the "Build hooks" section on your site dashboard at `/settings/deploys#build-hooks`).

Next, add the URL to your site .env variables and activate the plugin. Env variables are included in `.env.example` and below, for your reference:

```bash
## Hooks
NETLIFY_WEBHOOK_DEVELOPMENT=https://api.netlify.com/build_hooks/{yourBuildHookId}
NETLIFY_WEBHOOK_STAGING=https://api.netlify.com/build_hooks/{yourBuildHookId}
NETLIFY_WEBHOOK_PRODUCTION=https://api.netlify.com/build_hooks/{yourBuildHookId}
```

### Filters

#### netlify_posttypes

By default the plugin makes a run on the provided Netlify webhook when the standard WordPress posttypes `post` and `page` undergo a change in `publish` status.

If you would like to modify this you can do so by passing an array of desired posttypes to the `netlify_posttypes` filter.

```php
add_filter('netlify_posttypes', [
  'post',
  'page',
  'video-film',
  'brandon-small-jokes',
]);
```

#### netlify_webhooks

You can modify your webhooks at runtime using the `netlify_hooks` filter:

```php
add_filter('netlify_hooks', [
  'development' => 'https://api.netlify.com/build_hooks/########',
  'testing'     => 'https://api.netlify.com/build_hooks/########',
  'production'  => 'https://api.netlify.com/build_hooks/########',
])
```

#### netlify_env_override

If you don't want to use env variables because you don't have a deployment strategy and enjoy living poorly you can hook into the `netlify_env_override` filter and pass the target webhook directly at runtime:

```php
add_filter('netlify_env_override', 'https://api.netlify.com/build_hooks/########');
```

#### netlify_transitions

Change the post status transitions which trigger a build. Usage with the default values is shown below:

```php
add_filter('netlify_transitions', [
  'draft_to_publish',
  'publish_to_draft',
  'publish_to_trash',
  'publish_to_private',
  'private_to_public',
  'new_to_publish',
]);
```

## Author

👤 **Tiny Pixel Collective, LLC**

* Twitter: [@tinydevteam](https://twitter.com/tinydevteam)
* Github: [@pixelcollective](https://github.com/pixelcollective)
* Web: [tinypixel.dev](https://tinypixel.dev)

## 🤝 Contributing

Contributions, issues and feature requests are welcome!

🧠 ? [Open a PR](https://github.com/pixelcollective/netlify-deploy/pulls). Be sure to [abide by our contribution guidelines](https://github.com/pixelcollective/netlify-deploy/tree/master/CONTRIBUTORS.md).

😩 ? [Open an issue](https://github.com/pixelcollective/netlify-deploy/issues).

## Show your support

We need ⭐️s to live. _Please_.

## 📝 License

This project is [MIT](https://github.com/pixelcollective/netlify-deploy/blob/master/LICENSE) licensed.

***

Copyright © 2019 [Tiny Pixel Collective, LLC](https://github.com/@pixelcollective).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.6] - 2019-06-25

### Added

- The entire project.

[Unreleased]: https://github.com/pixelcollective/netlify-deploy/compare/v0.0.6...HEAD
[0.0.6]: https://github.com/pixelcollective/netlify-deploy/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/pixelcollective/netlify-deploy/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/pixelcollective/netlify-deploy/compare/v0.0.3...v0.0.4
[0.0.3]: https://github.com/pixelcollective/netlify-deploy/compare/v0.0.2...v0.0.3
[0.0.2]: https://github.com/pixelcollective/netlify-deploy/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/pixelcollective/netlify-deploy/releases/tag/v0.0.1


NetlifyDeploy/Error.php
<?php

namespace TinyPixel\NetlifyDeploy;

use \TinyPixel\NetlifyDeployPlugin as Plugin;
use function \admin_url;

class Error
{
    /**
     * Construct
     * @param \TinyPixel\NetlifyDeployPlugin
     */
    public function __construct($plugin)
    {
        $this->plugin = $plugin;
    }

    /**
     * Deactivates plugin and displays errors
     * @param array $error
     */
    public static function throw($error = null)
    {
        // this runs too late for deactivation
        // so we just manually include the function
        require_once ABSPATH . 'wp-admin/includes/plugin.php';

        // if plugin is activated, deactivate it
        is_plugin_active(Plugin::getInstance()->name) &&
            deactivate_plugins(Plugin::getInstance()->name);

        // if error was set cast it to an object
        if (!is_null($error)) {
            $error = is_array($error) ? (object) $error : $error;
        }

        // get formatted error variables
        $message = (object) [
            'title'    => isset($error->title)    ? $error->title : self::defaultTitle(),
            'subtitle' => isset($error->subtitle) ? $error->subtitle : self::defaultSubtitle(),
            'body'     => isset($error->body)     ? $error->body : self::defaultBody(),
            'footer'   => isset($error->footer)   ? $error->footer : self::defaultFooter(),
            'link'     => isset($error->link)     ? $error->link : self::defaultLink(),
        ];

        // prepare the liturgy
        $dirge = sprintf(
            "<h1>%s<br><small>%s</small></h1><p>%s</p><p>%s</p>",
            $message->title,
            $message->subtitle,
            $message->body,
            $message->footer
        );

        // bear the bad news
        wp_die($dirge, $message->title, $message->subtitle);
    }

    /**
     * Get default error message title
     * @return i18n formatted string
     */
    public static function defaultTitle()
    {
        return __(
            'Netlify Deploy Runtime Error',
            'netlify-deploy',
        );
    }

    /**
     * Get default error message subtitle
     * @return i18n formatted string
     */
    public static function defaultSubtitle()
    {
        return __(
            'There is a problem with the plugin',
            'netlify-deploy',
        );
    }

    /**
     * Get default error message body
     * @return i18n formatted string
     */
    public static function defaultBody()
    {
        return __(
            'There was a problem with the plugin.',
            'netlify-deploy',
        );
    }

    /**
     * Get default error message footer
     * @return i18n formatted string
     */
    public static function defaultFooter()
    {
        return __(
            'The plugin has been deactivated.',
            'netlify-deploy'
        );
    }

    /**
     * Get default error message link
     * @return array link
     */
    public static function defaultLink()
    {
        return [
            'link_text' => __('Plugin Administration ⌫', 'netlify-deploy'),
            'link_url'  => admin_url('plugins.php')
        ];
    }
}


NetlifyDeployPlugin.php
<?php
/**
 * Plugin Name:     Netlify Deploy
 * Plugin URI:      https://github.com/pixelcollective/netlify-deploy
 * Description:     Run netlify builds when site content is updated
 * Version:         0.1.0
 * Author:          Tiny Pixel Collective
 * Author URI:      https://tinypixel.dev
 * License:         MIT License
 * Text Domain:     netlify-deploy
 * Domain Path:     /languages
 */

namespace TinyPixel;

class NetlifyDeployPlugin
{
    /**
     * Instance
     * @static object
     */
    public static $instance;

    /**
     * Composer autoload location
     * @var string
     */
    public $dependencies = __DIR__ . '/vendor/autoload.php';

    /**
     * Error handler in case of autoload failure
     * @var string
     */
    public $errorHandler = __DIR__ . '/src/Error.php';

    /**
     * Default PostTypes
     * @var array
     */
    public $postTypes = [
        'post',
        'page',
    ];

    /**
     * Containerized classes
     * @var array
     */
    public $services = [
        'netlify' => '\TinyPixel\NetlifyDeploy\Netlify',
        'error'   => '\TinyPixel\NetlifyDeploy\Error',
    ];

    /**
     * Setup class parameters
     */
    public function __construct()
    {
        /**
         * Plugin name
         * @var string
         */
        $this->name = basename(dirname(__FILE__));

        /**
         * Plugin service container
         * @var object
         */
        $this->container = (object) [];

        /**
         * Plugin requirements
         * @var object
         */
        $this->requires = (object) [
            'php'  => '7.2',
            'wp'   => '5.2',
        ];
        /**
         * Runtime
         * @var string
         */
        $this->runtime = (object) [
            'env' => strtoupper(env('WP_ENV')),
            'php' => phpversion(),
        ];

        /**
         * WordPress profile & references
         * @var object
         */
        $this->wordpressVersion = get_bloginfo('version');

        /**
         * Location of plugin translation files
         * @var string
         */
        $this->i18nHandle = "{$this->name}/languages";
    }

    /**
     * Returns singleton construct
     * @return self $instance
     */
    public static function getInstance()
    {
        if (self::$instance == null) {
            self::$instance = new NetlifyDeployPlugin();
        }

        return self::$instance;
    }

    /**
     * Does compatibility checks
     */
    public function preflight()
    {
        $this
            ->loadTextDomain()
            ->checkDependencies()
            ->checkPHPVersion()
            ->checkWPVersion()
            ->checkEnv();

        return $this;
    }

    /**
     * Executes plugin runtime
     */
    public function run()
    {
        $this->container->netlify
            ->setupClient()
            ->setWebhooks()
            ->filterPostTypes()
            ->filterWebhooks()
            ->filterTransitions()
            ->filterOverrides()
            ->setTargetHook()
            ->addActions();
    }

    /**
     * Checks autoload viability
     */
    private function checkDependencies()
    {
        // If there is an issue with the autoloader this manually
        // loads the error handler class and uses it to throw an early
        // error
        !file_exists($this->dependencies) && $this->throwEarlyError([
            'body'     => __('Netlify Deploy needs to be installed in order to be run.<br />Run <code>composer install</code> from the plugin directory.', 'netlify-deploy'),
            'subtitle' => __('Autoloader not found.', 'netlify-deploy'),
        ]);

        // load dependencies
        require $this->dependencies;

        // bind services to container
        $this->bindServices();

        return $this;
    }

    /**
     * Binds service classes
     */
    private function bindServices()
    {
        foreach ($this->services as $handle => $class) {
            $this->container->$handle = new $class(self::getInstance());
        }
    }

    /**
     * Loads plugin i18n texts
     * @return self $this
     */
    private function loadTextDomain()
    {
        load_plugin_textdomain(
            'netlify-deploy',
            false,
            $this->i18nHandle
        );

        return $this;
    }

    /**
     * Checks for minimum PHP version compatibility
     * @return self $this
     */
    private function checkPHPVersion()
    {
        version_compare($this->requires->php, $this->runtime->php, '>') &&
            $this->container->error::throw([
                'body' => sprintf(
                    __('You must be using PHP %s or greater.', 'netlify-deploy'),
                    $this->requires->php,
                ),
                'subtitle' => sprintf(
                    __('Invalid PHP version (%s)', 'netlify-deploy'),
                    $this->runtime->php,
                ),
            ]);

        return $this;
    }

    /**
     * Checks for minimum WordPress version compatibility
     * @return self $this
     */
    private function checkWPVersion()
    {
        version_compare($this->requires->wp, $this->wordpressVersion, '>') &&
            $this->container->error::throw([
                'body' => sprintf(__(
                    'You must be using WordPress %s or greater',
                    'netlify-deploy'
                ), $this->requires->wp),
                'subtitle' => sprintf(
                    __('Invalid WordPress version (%s)', 'netlify-deploy'),
                    $this->wordpressVersion,
                ),
            ]);

        return $this;
    }

    /**
     * Checks environment variables
     * @return self $this
     */
    private function checkEnv()
    {
        $webhook = $this->container->netlify::getNetlifyHook(
            $this->runtime->env
        );

        if (!$webhook->value && !(has_filter('netlify_webhooks') || has_filter('netlify_env_override'))) {
            $this->container->error::throw([
                    'body' => sprintf(__(
                        "The <code>%s</code> variable must be present.",
                        'netlify-deploy'
                    ), $webhook->key, $this->runtime->env),
                    'subtitle' => __('Netlify webhook not found.', 'netlify-deploy'),
                ]);
        }

        return $this;
    }

    /**
     * Throws an error before autoload
     * @param array $error
     */
    private function throwEarlyError($error)
    {
        require $this->errorHandler;

        \TinyPixel\NetlifyDeploy\Error::throw([
            'body'     => __('Netlify Deploy needs to be installed in order to be run.<br />Run <code>composer install</code> from the plugin directory.', 'netlify-deploy'),
            'subtitle' => __('Autoloader not found.', 'netlify-deploy'),
        ]);
    }
}

NetlifyDeployPlugin
    ::getInstance()
    ->preflight()
    ->run();


NetlifyDeploy/Netlify.php
<?php

namespace TinyPixel\NetlifyDeploy;

use \GuzzleHttp\Client;

class Netlify
{
    /**
     * Construct
     */
    public function __construct($plugin)
    {
        $this->plugin = $plugin;

        /**
         * Post Status Transitions
         */
        $this->transitions = [
            'draft_to_publish',
            'publish_to_draft',
            'publish_to_trash',
            'publish_to_private',
            'private_to_public',
            'new_to_publish',
        ];


        return $this;
    }

    /**
     * Initializes http client for webhook request
     * @return self $this
     */
    public function setupClient()
    {
        // instantiate guzzle for POST req
        $this->client = new Client(['headers' => [
            'Content-Type' => 'application/json',
        ]]);

        return $this;
    }

    /**
     * Hooks into WordPress lifecycle
     * @return self $this
     */
    public function addActions()
    {
        foreach ($this->transitions as $transition) {
            add_action($transition, [
                $this, 'onStatusTransition'
            ], 10, 3);
        }

        foreach ($this->postTypes as $postType) {
            add_action("publish_{$postType}", [
                $this, 'postToWebhook',
            ], 10, 2);
        }

        return $this;
    }

    /**
     * On Status Transition
     * @param WP_Post $post
     */
    public function onStatusTransition($post)
    {
        if (in_array($post->post_type, $this->postTypes)) {
            $this->postToWebhook();
        }
    }

    /**
     * POSTs to appropriate webhook on publish actions
     */
    public function postToWebhook()
    {
        // make the run
        $this->targetHook &&
            $this->client->post($this->targetHook);
    }

    /**
     * Sets webhook URLs for POSTing
     * @return self $this
     */
    public function setWebhooks()
    {
        $this->hooks = (object) [
            'development' => self::getNetlifyHook('DEVELOPMENT')->value,
            'staging'     => self::getNetlifyHook('STAGING')->value,
            'production'  => self::getNetlifyHook('PRODUCTION')->value,
        ];

        return $this;
    }

    /**
     * Return hook for a given environment
     * @param string $environment
     * @return object
     */
    public static function getNetlifyHook($environment)
    {
        return (object) [
            'key'   => "NETLIFY_WEBHOOK_{$environment}",
            'value' => env("NETLIFY_WEBHOOK_{$environment}"),
        ];
    }

    /**
     * Filters webhooks at runtime
     * @return self $this
     */
    public function filterWebhooks()
    {
        has_filter('netlify_webhooks') &&
            apply_filters('netlify_webhooks', $this->hooks);

        return $this;
    }

    /**
     * Filters envvar overrides
     * @return self $this
     */
    public function filterOverrides()
    {
        $this->override = '';

        has_filter('netlify_env_override') &&
            apply_filters('netlify_env_override', $this->override);

        return $this;
    }

    /**
     * Fields PostTypes that are set to trigger webhook
     * @param array postTypes
     * @return self $this
     */
    public function filterPostTypes()
    {
        has_filter('netlify_posttypes') &&
            apply_filters('netlify_posttypes', $this->plugin->postTypes);

        $this->postTypes = $this->plugin->postTypes;

        return $this;
    }

    /**
     * Post status transitions
     * @var array
     */
    public function filterTransitions()
    {
        has_filter('netlify_transitions') &&
            apply_filters('netlify_transitions', $this->transitions);

        return $this;
    }

    /**
     * Set target hook, accounting for overrides
     * @return self $this
     */
    public function setTargetHook()
    {
        if ($this->override !== '') {
            $this->targetHook = $this->override;
        } elseif ($this->plugin->runtime->env) {
            $this->targetHook = self::getNetlifyHook(
                $this->plugin->runtime->env
            )->value;
        }

        return $this;
    }
}


netlify-deploy.php
<?php
/**
 * Plugin Name:     Netlify Deploy
 * Plugin URI:      https://github.com/pixelcollective/netlify-deploy
 * Description:     Run netlify builds when site content is updated
 * Version:         0.1.0
 * Author:          Tiny Pixel Collective
 * Author URI:      https://tinypixel.dev
 * License:         MIT License
 * Text Domain:     netlify-deploy
 * Domain Path:     /languages
 */

namespace TinyPixel;

(new class {

    /**
     * Base string for envvars
     * @var string
     */
    public $netlifyEnv = 'NETLIFY_WEBHOOK_';

    /**
     * Composer autoload location
     * @var string
     */
    public $composer = __DIR__ . '/vendor/autoload.php';

    /**
     * PostTypes that result in called webhook
     * @var array
     */
    public $postTypes = [
        'post',
        'page',
    ];

    /**
     * Setup class parameters
     */
    public function __construct()
    {
        /**
         * Environmental variables
         * @var string
         */
        $this->currentEnv = strtoupper(env('WP_ENV'));

        /**
         * Server PHP profile
         * @var string
         */
        $this->php = phpversion();

        /**
         * Plugin profile & references
         * @var object
         */
        $this->plugin = (object) [
            'ref'      => plugin_basename(__FILE__),
            'name'     => __('Netlify Deploy', 'netlify-deploy'),
            'author'   => __('Tiny Pixel Collective, LLC', 'netlify-deploy'),
            'link'     => 'https://tinypixel.dev/netlify-deploy',
            'license'  => 'MIT',
            'requires' => (object) [
                'php'  => '7.2',
                'wp'   => '5.2',
            ],
            'translationDir' => basename(dirname(__FILE__)) . '/languages',
        ];

        /**
         * WordPress profile & references
         * @var object
         */
        $this->wp = (object) [
            'version'     => get_bloginfo('version'),
            'adminLinks'  => (object) [
                'plugins' => admin_url('plugins.php'),
            ],
            'lib' => (object) [
                'plugins' => ABSPATH . 'wp-admin/includes/plugin.php',
            ],
        ];

        /**
         * Error message template string defaults
         * @var object
         */
        $this->errorDefaults = (object) [
            'title'    => __('Netlify Deploy Runtime Error', 'netlify-deploy'),
            'subtitle' => '',
            'body'     => __('There was a problem with the plugin.', 'netlify-deploy'),
            'footer'   => __('The plugin has been deactivated.', 'netlify-deploy'),
            'link' => [
                'link_text' => __('Plugin Administration ⌫', 'netlify-deploy'),
                'link_url'  => $this->wp->adminLinks->plugins,
            ],
        ];
    }

    /**
     * Plugin runtime
     */
    public function run()
    {
        // preflight
        $this->loadTextDomain()
                ->checkPHPVersion()
                ->checkWPVersion()
                ->checkEnv()
                ->checkComposer();

        // business
        $this->setupClient()
                ->setupNetlifyWebhooks()
                ->hookPublish();
    }

    /**
     * Initializes http client for webhook request
     * @return self $this
     */
    public function setupClient()
    {
        // load dependencies
        require $this->composer;

        // instantiate guzzle for POST req
        $this->client = new \GuzzleHttp\Client(['headers' => [
            'Content-Type' => 'application/json',
        ]]);

        return $this;
    }

    /**
     * Sets webhook URLs for POSTing
     * @return self $this
     */
    private function setupNetlifyWebhooks()
    {
        $this->netlify = (object) [
            'hooks' => (object) [
                'development' => env("{$this->netlifyEnv}DEVELOPMENT"),
                'staging'     => env("{$this->netlifyEnv}STAGING"),
                'production'  => env("{$this->netlifyEnv}PRODUCTION"),
            ],
        ];

        has_filter('netlify_webhooks') &&
            apply_filters('netlify_webhooks', $this->netlify->hooks);

        return $this;
    }

    /**
     * Hooks into WordPress lifecycle
     */
    private function hookPublish()
    {
        // mile high
        foreach ($this->getPostTypes() as $type) {
            add_action("publish_{$type}", [
                $this, 'onPublish'
            ], 10, 2);
        }
    }

    /**
     * Fields PostTypes that are set to trigger webhook
     * @return array $postTypes
     */
    private function getPostTypes()
    {
        has_filter('netlify_posttypes') &&
            apply_filters('netlify_posttypes', $this->postTypes);

        return $this->postTypes;
    }

    /**
     * POSTs to appropriate webhook on publish actions
     */
    public function onPublish()
    {
        // set the netlify hook based on envvar
        switch ($this->currentEnv) :
            case 'DEVELOPMENT':
                $this->hook = $this->netlify->hooks->development;
                break;

            case 'STAGING':
                $this->hook = $this->netlify->hooks->staging;
                break;

            case 'PRODUCTION':
                $this->hook = $this->netlify->hooks->production;
                break;
        endswitch;

        // make the run
        $this->hook &&
            $this->client->post($this->hook);
    }

    /**
     * Checks for minimum PHP version compatibility
     * @return self $this
     */
    private function checkPHPVersion()
    {
        version_compare($this->plugin->requires->php, $this->php, '>') && $this->error([
            'body' => sprintf(
                __('You must be using PHP %s or greater.', 'netlify-deploy'),
                $this->plugin->requires->php,
            ),
            'subtitle' => sprintf(
                __('Invalid PHP version (%s)', 'netlify-deploy'),
                $this->php,
            ),
        ]);

        return $this;
    }

    /**
     * Checks for minimum WordPress version compatibility
     * @return self $this
     */
    private function checkWPVersion()
    {
        version_compare($this->plugin->requires->wp, $this->wp->version, '>') && $this->error([
            'body' => sprintf(
                __('You must be using WordPress %s or greater', 'netlify-deploy'),
                $this->plugin->requires->wp,
            ),
            'subtitle' => sprintf(
                __('Invalid WordPress version (%s)', 'netlify-deploy'),
                $this->wp->version,
            ),
        ]);

        return $this;
    }

    /**
     * Checks for vendor dependencies
     * @return self $this
     */
    private function checkComposer()
    {
        !file_exists($this->composer) && $this->error([
            'body'     => __('Netlify Deploy needs to be installed in order to be run.<br />Run <code>composer install</code> from the plugin directory.', 'netlify-deploy'),
            'subtitle' => __('Autoloader not found.', 'netlify-deploy'),
        ]);

        return $this;
    }

    /**
     * Checks for environment variables
     * @return self $this
     */
    private function checkEnv()
    {
        !env("{$this->netlifyEnv}{$this->currentEnv}") && $this->error([
            'body' => sprintf(
                __("The <code>%s%s</code> variable must be present.", 'netlify-deploy'),
                $this->netlifyEnv,
                $this->currentEnv
            ),
            'subtitle' => __('Netlify webhook not found.', 'netlify-deploy'),
        ]);

        return $this;
    }

    /**
     * Load plugin translations
     */
    private function loadTextDomain()
    {
        load_plugin_textdomain('netlify-deploy', false, $this->plugin->translationDir);

        return $this;
    }

    /**
     * Handles deactivating plugin and displaying errors
     * @param array $error
     */
    private function error($error)
    {
        // deactivate self
        $this->forceEjectPlugin();

        // get formatted error variables
        $dirge = $this->processError($error);

        // bear the bad news
        wp_die(
            $dirge->message,
            $dirge->title,
            $dirge->link,
        );
    }

    /**
     * Immediately deactivate plugin regardless of WordPress lifecycle
     */
    private function forceEjectPlugin()
    {
        // this runs too late for deactivation
        // so we just manually include the function
        require_once $this->wp->lib->plugins;

        is_plugin_active($this->plugin->ref) &&
            deactivate_plugins($this->plugin->ref);
    }

    /**
     * Processes error message for display to user
     *
     * @param array $error
     * @return object $error
     */
    private function processError($error)
    {
        $error = is_array($error) ? (object) $error : $error;

        $errorObj = (object) [
            'title'    => $this->errorTitle($error),
            'subtitle' => $this->errorSubtitle($error),
            'body'     => $this->errorBody($error),
            'footer'   => $this->errorFooter($error),
            'link'     => $this->errorLink($error),
        ];

        $errorObj->message = $this->errorMessage($errorObj);

        return $errorObj;
    }

    /**
     * Formats error message
     * @return string
     */
    private function errorMessage($error)
    {
        return sprintf(
            "<h1>%s<br><small>%s</small></h1><p>%s</p><p>%s</p>",
            $error->title,
            $error->subtitle,
            $error->body,
            $error->footer,
        );
    }

    /**
     * Returns error title from error object
     * @return string
     */
    private function errorTitle($error)
    {
        return isset($error->title)
            ? $error->title
            : $this->errorDefaults->title;
    }

    /**
     * Returns error subtitle from error object
     * @return string
     */
    private function errorSubtitle($error)
    {
        return isset($error->subtitle)
            ? $error->subtitle
            : $this->errorDefaults->subtitle;
    }

    /**
     * Returns error body from error object
     * @return string
     */
    private function errorBody($error)
    {
        return isset($error->body)
            ? $error->body
            : $this->errorDefaults->body;
    }

    /**
     * Returns error link from error object
     * @return string
     */
    private function errorLink($error)
    {
        return isset($error->link)
            ? $error->link
            : $this->errorDefaults->link;
    }

    /**
     * Returns footer from error object
     * @return string
     */
    private function errorFooter($error)
    {
        return isset($error->footer)
            ? $error->footer
            : $this->errorDefaults->footer;
    }
})->run();
